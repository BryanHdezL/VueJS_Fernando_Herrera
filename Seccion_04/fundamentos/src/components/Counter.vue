<template>
  <h2>{{ customTitle }}</h2>
  <p>{{ counter }} <sup>2</sup> = {{ squareCounter }}</p>
  <div>
    <button @click="increase">+1</button>
    <button @click="decrease">-1</button>
  </div>
</template>

<script>
  export default {
    // Propiedades
    props: {
      title: {
        type: String,
      },
      start: {
        type: Number,
        default: 100,
        validator(value) {
          return value > 0;
        },
      },
    },
    // Datos
    data() {
      return {
        counter: 5,
      };
    },
    // Metodos
    methods: {
      //   Get Square Counter
      getSquareValue() {
        return this.counter * this.counter;
      },
      //   Increase
      increase() {
        this.counter++;
      },
      //   Decrease
      decrease() {
        this.counter--;
      },
    },
    // Cache
    computed: {
      //   Square Counter
      squareCounter() {
        return this.counter * this.counter;
      },
      //   Custom Title
      customTitle() {
        return this.title ? this.title : "Counter";
      },
    },
  };
</script>

<style>
  button {
    background: #64bb87;
    border-radius: 5px;
    border: 1px solid #ffffff;
    color: #ffffff;
    cursor: pointer;
    margin: 0 5px;
    padding: 5px 15px;
    transition: 0.3s ease-in-out;
  }
  button:hover {
    background: #5aa67b;
    transition: 0.3s ease-in-out;
  }
</style>
