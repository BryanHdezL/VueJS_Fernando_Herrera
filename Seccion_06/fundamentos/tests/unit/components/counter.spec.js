// Importacion de componente
import { shallowMount, mount } from '@vue/test-utils'
import Counter from '@/components/Counter'

// Test del componente Counter
describe('Counter Component', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Counter)
  })

  // Prueba para comporbar que el componente no se modifico
  test('Debe de hacer match con el snapshot', () => {
    // Evaluamos el HTML del componente
    expect(wrapper.html()).toMatchSnapshot()
  })

  // Prueba para verificar el valor de una etiqueta HTML
  test('H2 debe de terner el valor defecto "Counter"', () => {
    const wrapper = shallowMount(Counter)
    expect(wrapper.find('h2').exists()).toBeTruthy()
    const h2Value = wrapper.find('h2').text()
    expect(h2Value).toBe('Counter');
  })

  // Prueba con find & findAll
  test('El valor defecto debe de ser 100 en el parrafo', () => {
    // const pTags = wrapper.findAll('p')
    // expect(pTags[1].text()).toBe('100')
    const value = wrapper.find('[data-testid="counter"]').text()
    expect(value).toBe('100')
    expect(value).toBe('100')
  })

  // Prueba de eventos
  test('Debe de incrementar y decremenetar el contador', async () => {
    const [increaseBtn, decreaseBtn] = wrapper.findAll('button')
    await increaseBtn.trigger('click')
    await increaseBtn.trigger('click')
    await increaseBtn.trigger('click')
    await decreaseBtn.trigger('click')
    await decreaseBtn.trigger('click')
    const value = wrapper.find('[data-testid="counter"]').text()
    expect(value).toBe('101')
  })

  // Prueba de propiedades
  test('Debe de establecer el valor por defecto', () => {
    const { start } = wrapper.props()
    const value = wrapper.find('[data-testid="counter"]').text()
    expect(Number(value)).toBe(start)
  });

  // Prueba de propiedades enviadas
  test('Debe de mostrar la propiedad title', () => {
    const wrapper = shallowMount(Counter, {
      props: {
        title: 'Hola Mundo'
      }
    })
    expect(wrapper.find('h2').text()).toBe('Hola Mundo')
  });

})