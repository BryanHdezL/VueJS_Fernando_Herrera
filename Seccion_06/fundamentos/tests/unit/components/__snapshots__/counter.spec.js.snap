// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component Debe de hacer match con el snapshot 1`] = `
<h2>Counter</h2>
<p>100 <sup>2</sup> = 10000</p>
<p data-testid="counter">100</p>
<div class="buttons-container"><button>+1</button><button>-1</button></div>
`;
